ccl/sqlproxyccl: provide hints to the user whenever params parsing fails

jaylim-crl opened the pull request:
Previously, whenever params parsing fails during connection startup, an error
message is returned without any hints. Users often get confused on what to
do next as the message can be ambiguous and provide no action items. This
led to a lot of confusion among end users connecting through the sqlproxy,
which can be seen in the recent posts in community forums.

This commit updates the error messages to also return hints, and provide users
with more information. At the same time, we also relabeled "cluster name" to
"cluster identifier", which correctly reflects the format that we're using at
the moment: <cluster name>-<tenant ID>.

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: provide hints to the user whenever params parsing fails

Previously, whenever params parsing fails during connection startup, an error
message is returned without any hints. Users often get confused on what to
do next as the message can be ambiguous and provide no action items. This
led to a lot of confusion among end users connecting through the sqlproxy,
which can be seen in the recent posts in community forums.

This commit updates the error messages to also return hints, and provide users
with more information. At the same time, we also relabeled "cluster name" to
"cluster identifier", which correctly reflects the format that we're using at
the moment: <cluster name>-<tenant ID>.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72277)
<!-- Reviewable:end -->

jaylim-crl reviewed (commented):
With this PR, we'll provide hints whenever a user encounters an error message related to cluster identifiers when connecting through the proxy. Here's an example:

![missing-name.png](https://files.reviewableusercontent.io/images/reviewable/83607639/WXAWjjApXy33c7J_TaWQUa6k/missing-name.png)

cc: @emily-horing @jordanlewis 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-:-MnRTim97XXsEC0HwZ1F:b-yzaoyn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @JeffSwenson, and @kernfeld-cockroach)



<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, and @kernfeld-cockroach)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRgs0n2MiRjUoM0Tq4:-MnRgs0n2MiRjUoM0Tq5:bmr7raj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy.go#L63)):*
> ```Go
> 			Message:  msg,
> 		}
> 		hint := errors.FlattenHints(codeErr.err)
> ```

nit: This could be simplified to
resp := &pgproto3.ErrorResponse {
	Severity: "FATAL",
	Code:     pgCode,
	Message:  msg,
	Hint: errors.FlattenHints(codeErr.err),
}
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 742 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRiDsL495ceczXNChg:-MnRiDsL495ceczXNChh:b-pm6a9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy_handler.go#L742)):*
> ```Go
> const clusterIdentifierHint = `
> ensure that your cluster identifier is specified through only one of the following methods:
> ```

I like to link to documentation when an error has a clear doc. What do you think about including a link to:

https://www.cockroachlabs.com/docs/cockroachcloud/connect-to-a-serverless-cluster


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-:-MnRjhWEAi4SwH7dc3AJ:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @JeffSwenson, and @kernfeld-cockroach)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRgs0n2MiRjUoM0Tq4:-MnRjYpz0flOtJxReOS5:b-h0j0w9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy.go#L63)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: This could be simplified to
resp := &pgproto3.ErrorResponse {
	Severity: "FATAL",
	Code:     pgCode,
	Message:  msg,
	Hint: errors.FlattenHints(codeErr.err),
}
</blockquote></details>

Thanks. Will update.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 742 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRiDsL495ceczXNChg:-MnRj_DLD2IjFWnAOXfp:b-orwsbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy_handler.go#L742)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

I like to link to documentation when an error has a clear doc. What do you think about including a link to:

https://www.cockroachlabs.com/docs/cockroachcloud/connect-to-a-serverless-cluster
</blockquote></details>

The problem is that this sqlproxy isn't specific to CC. Users could also use it for their own multi-tenant setup if they wanted to. If we linked a doc to cockroachcloud, that would be incorrect.


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @JeffSwenson, and @kernfeld-cockroach)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRgs0n2MiRjUoM0Tq4:-MnRkbIR8eKJv5Fb7WZP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy.go#L63)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

Thanks. Will update.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, @JeffSwenson, and @kernfeld-cockroach)

___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 742 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRiDsL495ceczXNChg:-MnSHoA80JMPMhLZWq3j:b670i8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy_handler.go#L742)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

The problem is that this sqlproxy isn't specific to CC. Users could also use it for their own multi-tenant setup if they wanted to. If we linked a doc to cockroachcloud, that would be incorrect.
</blockquote></details>

I agree that we should link to the cloud docs. It's not reasonable to support all use cases, and we're certainly not expecting many people will use SQLProxy in their own systems - no need to worry about that case.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnSI5qdDoAgclpqzRrs:-MnSI5qdDoAgclpqzRrt:b-sxefok) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7dd17c0056f6553c7bcecbf4027cb3b24eef8f/pkg/ccl/sqlproxyccl/proxy_handler.go#L741)):*
> ```Go
> 
> const clusterIdentifierHint = `
> ensure that your cluster identifier is specified through only one of the following methods:
> ```

is it really *only* one? is there a problem using both?
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 743 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnSHxLuEzS73OkYU9km:-MnSHxLuEzS73OkYU9kn:b-n000t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7dd17c0056f6553c7bcecbf4027cb3b24eef8f/pkg/ccl/sqlproxyccl/proxy_handler.go#L743)):*
> ```Go
> ensure that your cluster identifier is specified through only one of the following methods:
> 
> 1) database parameter:
> ```

nit: let's capitalize Database parameter
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 747 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnSI-zPA348jkeGjsq-:-MnSI-zPA348jkeGjsq0:b8pe7ne) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7dd17c0056f6553c7bcecbf4027cb3b24eef8f/pkg/ccl/sqlproxyccl/proxy_handler.go#L747)):*
> ```Go
>    (e.g. "active-roach-42.defaultdb")
> 
> 2) options parameter:
> ```

nit: let's capitalize Oatabase parameter


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
TFTR! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-:-MnSJmIjA_vjZ2fFaePR:bz5m8k4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, @JeffSwenson, @jordanlewis, and @kernfeld-cockroach)

___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 742 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRiDsL495ceczXNChg:-MnSJKAuCcivQqVS_0Jn:b44x3ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy_handler.go#L742)):*
> we're certainly not expecting many people will use SQLProxy in their own systems 

Agreed. I'll update this accordingly. Also, the docs are slightly outdated; it references cluster-id, which is incorrect: 
![image.png](https://files.reviewableusercontent.io/images/reviewable/83607639/fmHB5MmmbAQSq3EmPQxl3ak1/image.png)

I'll sort that out too. 
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnSI5qdDoAgclpqzRrs:-MnSIuHo3tM2--HC9bkC:bbw5gkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7dd17c0056f6553c7bcecbf4027cb3b24eef8f/pkg/ccl/sqlproxyccl/proxy_handler.go#L741)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

is it really *only* one? is there a problem using both?
</blockquote></details>

Yes. If we specified both today, we'll get an error: https://github.com/cockroachdb/cockroach/blob/199c565e31a129f0c90d9fbd710feb852f5b1d93/pkg/ccl/sqlproxyccl/proxy_handler.go#L595-L597

There are no issues with using both. We can change that logic, and add a hint if a user specifies two different cluster names. I think back then we enforced this restriction to keep it simple.


<!-- Sent from Reviewable.io -->

jaylim-crl committed:
ccl/sqlproxyccl: allow multiple cluster identifiers provided they are unique

Previously, the proxy will return an error if multiple cluster identifiers
were specified in both the database and options parameters, regardless of
whether they are different or unique. Ideally we should only return an error
if the supplied cluster identifiers are non-ambiguous. We went with the
constraint earlier to simplify things, but thinking about it again, it seems
fine to specify cluster identifiers through both parameters provided that they
are unique.

This commit changes the behavior of the proxy such that we now allow multiple
cluster identifiers to be supplied within the connection string provided that
they are unique so that the proxy knows which cluster the connection is for.

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: relabel variables and comments to reflect cluster identifiers

Previously, we used the term "cluster name" in two different ways, one with
the tenant ID, and one without. This can be really confusing to readers who
are not familiar with the sqlproxy code. Concretely, cluster name is actually
the string without the tenant ID. The combination of cluster name and tenant
ID is known as the cluster identifier.

This commit is purely mechanical to fix comments and relabel variables.

Release note: None

jaylim-crl reviewed (commented):
Done. I have addressed all concerns. TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-:-MnTnA70BnlFqwniex2c:bl1ksqf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @JeffSwenson, @jordanlewis, and @kernfeld-cockroach)

___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 742 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnRiDsL495ceczXNChg:-MnTmJCAEvUjq4KOrODe:b-79kr6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0646354438196a5868f32db353dc0af2c658d315/pkg/ccl/sqlproxyccl/proxy_handler.go#L742)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

> we're certainly not expecting many people will use SQLProxy in their own systems 

Agreed. I'll update this accordingly. Also, the docs are slightly outdated; it references cluster-id, which is incorrect: 
![image.png](https://files.reviewableusercontent.io/images/reviewable/83607639/fmHB5MmmbAQSq3EmPQxl3ak1/image.png)

I'll sort that out too. 
</blockquote></details>

Done. I updated the hint to include the URL that Jeff linked:

```
#
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
FATAL: codeParamsRoutingFailed: missing cluster identifier in connection string
SQLSTATE: 08004
HINT: Ensure that your cluster identifier is uniquely specified using any of the
following methods:

1) Database parameter:
   Use "<cluster identifier>.<database name>" as your database parameter.
   (e.g. "postgres://.../active-roach-42.defaultdb?sslmode=...")

2) Options parameter:
   Use "--cluster=<cluster identifier>" as the options parameter.
   (e.g. "postgres://...&options=--cluster=active-roach-42")

For more details, please visit our docs site at:
   https://www.cockroachlabs.com/docs/cockroachcloud/connect-to-a-serverless-cluster
Failed running "sql"

```
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnSI5qdDoAgclpqzRrs:-MnTmUZ01MYNbP1Wtfcn:bi8zb1t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7dd17c0056f6553c7bcecbf4027cb3b24eef8f/pkg/ccl/sqlproxyccl/proxy_handler.go#L741)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

Yes. If we specified both today, we'll get an error: https://github.com/cockroachdb/cockroach/blob/199c565e31a129f0c90d9fbd710feb852f5b1d93/pkg/ccl/sqlproxyccl/proxy_handler.go#L595-L597

There are no issues with using both. We can change that logic, and add a hint if a user specifies two different cluster names. I think back then we enforced this restriction to keep it simple.
</blockquote></details>

Done. There won't be issues if we supplied the identifier through both mechanisms now, provided that they are non-ambiguous. However, if a user did something like options="--cluster=X --cluster=Y -c cluster=Z", the proxy will return an error. We can only specify one cluster key within the options parameter. We enforced this constraint for simplicity since we didn't want to parse the entire options string, and note that this is already the existing behavior of the proxy.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 743 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnSHxLuEzS73OkYU9km:-MnTmAKj7_AnM1o1g8Y7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7dd17c0056f6553c7bcecbf4027cb3b24eef8f/pkg/ccl/sqlproxyccl/proxy_handler.go#L743)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: let's capitalize Database parameter
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 747 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnSI-zPA348jkeGjsq-:-MnTmAtx-Jb6ylh3JWch:b-slhr3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7dd17c0056f6553c7bcecbf4027cb3b24eef8f/pkg/ccl/sqlproxyccl/proxy_handler.go#L747)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: let's capitalize Oatabase parameter
</blockquote></details>

Done. 
(I like the term Oatabase :smile:)


<!-- Sent from Reviewable.io -->

jordanlewis commented:
@rafiss could you PTAL at this updated info?

rafiss reviewed (commented):
i just had style comments. the wording lgtm, and thank you!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-:-MnXJX109mi2Yzj_cKYW:bousp1e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, @JeffSwenson, @jordanlewis, and @kernfeld-cockroach)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 53 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXHgNy9yUmcrNbX1Fr:-MnXHgNy9yUmcrNbX1Fs:b-u5scp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy.go#L53)):*
> ```Go
> 		var pgCode string
> 		if codeErr.code == codeIdleDisconnect {
> 			pgCode = "57P01" // admin shutdown
> ```

nit: change to `pgCode = pgcode.AdminShutdown`
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 55 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXHZyKA2a9YceRB0ag:-MnXHZyKA2a9YceRB0ah:bojs9f3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy.go#L55)):*
> ```Go
> 			pgCode = "57P01" // admin shutdown
> 		} else {
> 			pgCode = "08004" // rejected connection
> ```

nit: change to `pgCode = pgcode.SQLserverRejectedEstablishmentOfSQLconnection`
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 592 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXFsng0Zi9XHZ2pCPF:-MnXFsng0Zi9XHZ2pCPG:b-6udt5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy_handler.go#L592)):*
> ```Go
> 
> 	if clusterNameFromDB == "" && clusterNameFromOpt == "" {
> 		err := errors.New("missing cluster identifier")
> ```

optional nit: all of the errors.New could become
```
pgerror.New(pgcode.SQLserverRejectedEstablishmentOfSQLconnection, "...")
```

(or `pgerror.Newf`)

then you wouldn't need the extra pgCode logic in proxy.go
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 740 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXItTJ0FC1ZnG3LkWo:-MnXItTJ0FC1ZnG3LkWp:bu3lw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy_handler.go#L740)):*
> ```Go
> }
> 
> const clusterIdentifierHint = `
> ```

nit: i think it would be nicer to avoid the extra newline here, so we can remove all the `strings.TrimLeft` calls
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 757 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXGh9K83ir-ytCaKpk:-MnXGh9K83ir-ytCaKpl:b-97suep) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy_handler.go#L757)):*
> ```Go
> const clusterNameFormHint = "Cluster identifiers come in the form of <name>-<tenant ID> (e.g. lazy-roach-3)"
> 
> const missingTenantIDHint = "did you forget to include your tenant ID in the cluster identifier?"
> ```

nit: by convention, hints begin with upper case and use proper punctuation. same comment goes for the other hints above
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 601 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXJDAW4Qw-7VH5w3eH:-MnXJDAX4rVwqcQunyNY:blhrg75) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3cb5e36d9e1c5853a8aebbef952dfd3ec76cb3c/pkg/ccl/sqlproxyccl/proxy_handler.go#L601)):*
> ```Go
> 		err := errors.New("multiple different cluster identifiers provided")
> 		err = errors.WithHintf(err,
> 			"is '%s' or '%s' the identifier for the cluster that you're connecting to?",
> ```

nit: use upper case


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
TFTR! Addressed all of them except the optional nit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-:-MnXZsL_4MAUI4vvUKyK:bl4gkl6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @JeffSwenson, @jordanlewis, @kernfeld-cockroach, and @rafiss)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 53 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXHgNy9yUmcrNbX1Fr:-MnXZGvp6EFpuin_o2io:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy.go#L53)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: change to `pgCode = pgcode.AdminShutdown`
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 55 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXHZyKA2a9YceRB0ag:-MnXZHlr7wjTPF7xiZCk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy.go#L55)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: change to `pgCode = pgcode.SQLserverRejectedEstablishmentOfSQLconnection`
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 592 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXFsng0Zi9XHZ2pCPF:-MnXZWJL-bmLquBwZesG:b-xoba9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy_handler.go#L592)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

optional nit: all of the errors.New could become
```
pgerror.New(pgcode.SQLserverRejectedEstablishmentOfSQLconnection, "...")
```

(or `pgerror.Newf`)

then you wouldn't need the extra pgCode logic in proxy.go
</blockquote></details>

I left it as-is without this suggestion since the proxy only cares about two different types of codes: admin shutdown, and server rejected. If the proxy does end up needing more codes in the future, we can revisit this. I felt that `pgcode.SQLserverRejectedEstablishmentOfSQLconnection` was too long, and I don't think callers would need to know about it, a least for now.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 740 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXItTJ0FC1ZnG3LkWo:-MnXZO2v7BoEYo9NYY9m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy_handler.go#L740)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i think it would be nicer to avoid the extra newline here, so we can remove all the `strings.TrimLeft` calls
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 757 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXGh9K83ir-ytCaKpk:-MnXZRDA8zRTSRLCD79P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5281c779a703615a2c305b044bd1ecdf809ddb84/pkg/ccl/sqlproxyccl/proxy_handler.go#L757)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: by convention, hints begin with upper case and use proper punctuation. same comment goes for the other hints above
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy_handler.go, line 601 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/72277#-MnXJDAW4Qw-7VH5w3eH:-MnXZMozAqGOPO7H5Ysw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3cb5e36d9e1c5853a8aebbef952dfd3ec76cb3c/pkg/ccl/sqlproxyccl/proxy_handler.go#L601)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: use upper case
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
looks good from my end! thanks for the nice improvement

jaylim-crl commented:
TFTR!

bors r=JeffSwenson,rafiss
